### PR TITLE
[18 India] Create and Remove Gauge Change Markers 

### PR DESCRIPTION
**Pull request by EdPovi**
- [x] Branch is derived from the latest `master`
- [N/A] Add the `pins` or `archive_alpha_games` label if this change will break existing games
- [x] Code passes linter with `docker compose exec rack rubocop -a`
- [x] Tests pass cleanly with `docker compose exec rack rake`

### Explanation of Change
- Modified the `Part::Borders` class to add a `:gauge_change` type and it renders as a small circle on the edge. Added an `on_click` event that can be used to remove the border. 
- Added `RemoveGaugeChange` action, which stores the hex and edge of the removed border.
- Modified `remove_border_calculate_cost!` from `Tracker`, which modifies or deletes borders as track laying completes both edges of a border. 
- Created `Note` class to allow a message to be displayed.  This is used to display a note that provides instructions on how to remove a gauge change marker.
- The Note can be displayed on during an operating round by implementing `show_note?` in relevant step.
- The route distance is modified by checking number of borders crossed. (similar to 1848 code)

### Screenshots
image
image

### Any Assumptions / Hacks
- A dummy zero revenue `City` is inserted into the middle of the visited stops array.  This extends the route distance, but also ensures that the gauge changes isn't the first or last stop of the route. 
- A `border` is on both hexes of an edge, therefore the border count is halved for number of stops to add.

### CODE REVIEW

**Review by ollybh — CHANGES_REQUESTED** [review]
This generally looks reasonable to me, but the changes to the core code will need approval from someone like @tobymao, @michaeljb  or @crericha.
You will need to make sure that the game graph doesn't get recalculated during the game loading process, I've added a comment explaining this.

File: assets/app/view/game/note.rb
Comment (EdPovi) [reply assets/app/view/game/note.rb]: I tested it and we don't need the other respond_to? either.  I removed both. 

Side Note:  I also wonder if it would be useful to add the following to the `Step::Base` class, which would serve as a indication that the feature is available if others want to use it? 
```
def show_note?(_entity)
  false
end
def note_text_block
  []
end
```

File: lib/engine/game/g_18_india/step/track.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/step/track.rb:63]: Thanks for this note.  I learned a bit about this as I was working on walking a Track Laying route from a railhead.  

I agree with your suggestion to remove the `any_reachable_gauge_changes?(entity)` check from the method, as that is checked in the process call as you noted. 

The only downside is you may be shown the message that Gauge Change Markers may be removed, because they are on the map, but can't actually remove it as those markers are out of reach for the current corp. 

Thanks!

Comment (EdPovi) [reply lib/engine/game/g_18_india/step/track.rb:63]: Another idea would be to add a `return if @game.loading` guard clause in the `reachable_gauge_change?` method?

Comment (EdPovi) [reply lib/engine/game/g_18_india/step/track.rb:63]: Since `reachable_gauge_change?` is called from `process_remove_gauge_change`, which will also be called when running though all the actions on game load, I think the `game.loading` check will be more effective. 

Added the guard clause.

**Comment by EdPovi** [comment]
Thank you for your suggestions.  I have implemented them and should be ready for your review.

File: lib/engine/game/g_18_india/step/track.rb
Comment (ollybh) [reply lib/engine/game/g_18_india/step/track.rb:63]: No, that's the wrong way to do this.
The issue that you've now got is that `actions` will return different lists of possible actions when running interactively or non-interactively (game loading). That's potentially going to cause trouble.
I suspect that it probably wouldn't actually be an issue here, but it's something to be avoided. The problem could occur if `actions` returns an empty array when called from the game UI, then that player's turn would be skipped. If on game loading `actions` returns a non-empty array then their turn will not be skipped, and the game will break.
You want to move the `@game.loading` check to inside `process_remove_gauge_change`, so that the call to `reachable_gauge_change?` isn't made if `@game.loading` is true. That way the legality of the action is checked when it is first made, then when loading the game again we assume that the move is legal as it's been checked before getting added to the game history.

Comment (EdPovi) [reply lib/engine/game/g_18_india/step/track.rb:63]: I understand, thank you for the explanation.  I will adjust.

Comment (EdPovi) [reply lib/engine/game/g_18_india/step/track.rb]: Moved loading check.

Comment (ollybh) [reply lib/engine/game/g_18_india/step/track.rb]: > The only downside is you may be shown the message that Gauge Change Markers may be removed, because they are on the map, but can't actually remove it as those markers are out of reach for the current corp.

You can keep a graph check for reachable gauge change markers in `show_note?` as that method won't be called during the game load process.

**Review by ollybh — COMMENTED** [review]
I wonder if it would be better to make the name of the new action a bit more generic than `remove_gauge_change`? `remove_border` maybe? I don't know of any other games that have a similar mechanism, but I expect that if anything does need to reuse this functionality it might be for some other border edge action, not specifically a gauge change marker.

**Comment by EdPovi** [comment]
> I wonder if it would be better to make the name of the new action a bit more generic than `remove_gauge_change`? `remove_border` maybe? I don't know of any other games that have a similar mechanism, but I expect that if anything does need to reuse this functionality it might be for some other border edge action, not specifically a gauge change marker.

I named it to be clear in this context.  The gauge change is a type of border mostly because it is placed on an edge of a tile.  The only game I know of so far that has a gauge change border is 1848 Australia, which does behave similarly, but those markers are not removed.  I don't know of any games in development that plans to use a similar mechanic, but now that two have it, more may come later.  

I created the new action mostly because I needed to store two pieces of data to perform the action, a target Hex and the relevant Edge.  Calling it `remove_border ` is reasonable as that is the type of Part we modify. 

I have changed the name of the Action and tested that it works.

**Comment by EdPovi** [comment]
I have modified the Gauge Change Module so that it would be easier to add this module to 1848 for the visual change to Gauge Change Markers. 

In theory, this module could be added to 1848 without breaking games, but needs testing.

File: lib/engine/game/g_18_india/game.rb
Comment (ollybh) [reply lib/engine/game/g_18_india/game.rb:936]: It might be worth creating a GaugeChangeStopclass that's a subclass of Engine::Part::City that defaults to zero revenue.
If nothing else, it would let you rewrite this method as a single line:
```
route_stops.insert(1, *Array.new(num) { GaugeChangeStop.new })
```

File: lib/engine/game/g_18_india/step/gauge_change_border.rb
Comment (ollybh) [inline lib/engine/game/g_18_india/step/gauge_change_border.rb:34]: Is this the same as:
```suggestion
                neighbor.tile.borders.reject! { |nb| nb.edge == hex.invert(edge) }
```
Using `map!` to change a value to nil and then remove it with `compact!` feels a bit strange.

File: lib/engine/game/g_18_india/step/gauge_change_border.rb
Comment (ollybh) [inline lib/engine/game/g_18_india/step/gauge_change_border.rb:41]: As this code is within the G18India module you shouldn't be making it too generic and having `respond_to?` tests. If it gets extracted into shared code then this changes should be made. (And it would be better to have a method like `after_change_border` that gets called unconditionally, with a default implementation that's a no-op).

File: lib/engine/game/g_18_india/step/gauge_change_border.rb
Comment (ollybh) [inline lib/engine/game/g_18_india/step/gauge_change_border.rb:47]: This is repeating code from earlier in the method. Can it be moved so the repetition isn't needed, or extracted into a separate method?

File: lib/engine/game/g_18_india/step/track.rb
Comment (ollybh) [inline lib/engine/game/g_18_india/step/track.rb:91]: AFAICT, you're not calling the `remove_border_calculate_cost!` method you've added in the RemoveBorders module.

**Review by ollybh — CHANGES_REQUESTED** [review]

File: lib/engine/game/g_18_india/game.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/game.rb:931]: Removed check.

File: lib/engine/game/g_18_india/game.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/game.rb:936]: > What's this line for? Does this information end up being shown to the user in the route selector? It's going to be confusing if you have the same hex coordinates repeated several times.

The Hex isn't shown as part of the "route info" when selecting routes.  However, I needed it as there was another method that broke when the visited hexes had a `nil` value.

File: lib/engine/game/g_18_india/step/gauge_change_border.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/step/gauge_change_border.rb:34]: NOTE: This is in the existing code in `Engine::Step::Tracker` for `remove_border_calculate_cost!`.

I will test the change to see if there is a reason the existing cost wrote it this way.

File: lib/engine/game/g_18_india/step/gauge_change_border.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/step/gauge_change_border.rb:41]: That makes sense.  I removed the `respond_to?` test. 

Also, currently only 48 would be able to used the update, and it only needs to add them.  It doesn't need to remove them, so wouldn't need all of the elements.

File: lib/engine/game/g_18_india/step/gauge_change_border.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/step/gauge_change_border.rb:34]: Tested change, it works as expected.  Not sure why it was originally coded the other way.

File: lib/engine/game/g_18_india/step/gauge_change_border.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/step/gauge_change_border.rb:47]: extracted to separate method.

File: lib/engine/game/g_18_india/game.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/game.rb:934]: changing the class of the stop breaks other route checks.

Comment (EdPovi) [reply lib/engine/game/g_18_india/game.rb:934]: image

File: lib/engine/game/g_18_india/step/track.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/step/track.rb:91]: I can confirm that the `remove_border_calculate_cost!` method is called.  It is a part of every `lay_tile` action. 
image

**Comment by EdPovi** [comment]
Thank you for the review.  I have addressed your comments!

File: lib/engine/game/g_18_india/step/track.rb
Comment (ollybh) [reply lib/engine/game/g_18_india/step/track.rb:91]: Got it, I was confused with the two different places you had this code for removing borders, but one is for river costs and this one if for the gauge change markers. Right?

**Review by ollybh — APPROVED** [review]
I think this looks OK now, but I'd like @tobymao or @michaeljb to review the changes to the core code before it is merged.

File: lib/engine/game/g_18_india/step/track.rb
Comment (michaeljb) [inline lib/engine/game/g_18_india/step/track.rb:81]:
```suggestion
          # NOTE: Triggered by on_click event in View::Game::Part::Borders
```

File: lib/engine/game/g_18_india/step/track.rb
Comment (michaeljb) [inline lib/engine/game/g_18_india/step/track.rb:55]: See comments on assets/app/view/game/note.rb

Also, it is probably worth adding help/note text onto the running routes step, to indicate you don't need to click on the gauge change markers, but they still do count as stops.

**Review by michaeljb — CHANGES_REQUESTED** [review]
The main idea of the change looks good, I've mostly got nitpicky comments.

I don't think we need a new Note view, I think the Help view can be used instead and if there's anything lacking there we can change it https://github.com/tobymao/18xx/blob/master/assets/app/view/game/help.rb

File: assets/app/view/game/part/borders.rb
Comment (EdPovi) [reply assets/app/view/game/part/borders.rb:110]: That looks like a good change.  I copied this code from the method that already existed there, so we should change both.

File: lib/engine/game/g_18_india/game.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/game.rb:941]: I agree with the change in the method's name.  As a note, I have copied this name from 1848.

File: assets/app/view/game/note.rb
Comment (EdPovi) [reply assets/app/view/game/note.rb:1]: I wasn't aware of how `help` worked.  I have tested using the `help` view and it works well.  I have removed the Note view.

File: lib/engine/game/g_18_india/step/track.rb
Comment (EdPovi) [reply lib/engine/game/g_18_india/step/track.rb:57]: Converted this to use `help` view, it was straightforward to use help.  
Thanks for pointing out that class. 

I added help text to route step as suggested.

**Comment by EdPovi** [comment]
@michaeljb   Thank you for your review.  I have addressed all your comments and I have used the help class as you suggested.  

Thanks,
Ed

**Review by michaeljb — APPROVED** [review]

**Pull request merged**
